### PR TITLE
Improved AST memory footprint by calling ArrayList(optimal capacity);

### PR DESCRIPTION
**Pull request by leventov**
… fixed inconsistencies and possible bugs in several remove[Element]([Element] e) methods

### CODE REVIEW

**Comment by monperrus** [comment]
What's the actual memory gain for a some usage scenario?

**Comment by leventov** [comment]
Old footprint of Spoon project AST itself:
```
spoon.support.reflect.declaration.CtPackageImpl instance footprint:
 COUNT   AVG   SUM DESCRIPTION
     9  3674 33072 [B
 234368    47 11108720 [C
   326   522 170416 [I
     1  1952  1952 [J
 63493    56 3558000 [Ljava.lang.Object;
    13    46   600 [Ljava.lang.String;
     1    80    80 [Ljava.util.HashMap$Node;
     2    88   176 [Ljava.util.Hashtable$Entry;
     2    40    80 [Z
     1    24    24 com.intellij.rt.execution.junit.segments.SegmentedOutputStream
     1    24    24 java.io.BufferedOutputStream
     2    40    80 java.io.BufferedWriter
   319    32 10208 java.io.File
     2    40    80 java.io.FileDescriptor
     2    32    64 java.io.FileOutputStream
     4    24    96 java.io.OutputStreamWriter
     2    32    64 java.io.PrintStream
     2    16    32 java.lang.Boolean
    39    16   624 java.lang.Character
     1    96    96 java.lang.Class
     1    24    24 java.lang.Double
    16    16   256 java.lang.Integer
     5    24   120 java.lang.Long
     4    16    64 java.lang.Object
 234355    24 5624520 java.lang.String
     5    24   120 java.lang.StringBuffer
     1    24    24 java.math.RoundingMode
     4    48   192 java.nio.HeapByteBuffer
     1    16    16 java.nio.charset.CodingErrorAction
     2    64   128 java.text.DateFormatSymbols
     2   144   288 java.text.DecimalFormat
     2    64   128 java.text.DecimalFormatSymbols
     2    40    80 java.text.DigitList
     2    64   128 java.text.SimpleDateFormat
 63662    24 1527888 java.util.ArrayList
     1    24    24 java.util.Currency
     4    24    96 java.util.Date
     2   112   224 java.util.GregorianCalendar
   321    48 15408 java.util.HashMap
     1    32    32 java.util.HashMap$Node
   319    16  5104 java.util.HashSet
     2    48    96 java.util.Hashtable
    12    32   384 java.util.Hashtable$Entry
     1    32    32 java.util.Locale
     3    32    96 java.util.Stack
  9672    48 464256 java.util.TreeMap
  8009    40 320360 java.util.TreeMap$Entry
    60    16   960 java.util.TreeMap$KeySet
    20    16   320 java.util.TreeMap$Values
  9668    16 154688 java.util.TreeSet
     3    32    96 java.util.Vector
    12    24   288 org.apache.log4j.CategoryKey
     1    64    64 org.apache.log4j.ConsoleAppender
     1    16    16 org.apache.log4j.DefaultCategoryFactory
     1    48    48 org.apache.log4j.Hierarchy
     3    24    72 org.apache.log4j.Level
     6    40   240 org.apache.log4j.Logger
     2    32    64 org.apache.log4j.PatternLayout
     6    32   192 org.apache.log4j.ProvisionNode
     1    88    88 org.apache.log4j.RollingFileAppender
     2    16    32 org.apache.log4j.helpers.AppenderAttachableImpl
     1    40    40 org.apache.log4j.helpers.CountingQuietWriter
     2    24    48 org.apache.log4j.helpers.FormattingInfo
     2    24    48 org.apache.log4j.helpers.OnlyOnceErrorHandler
     2    48    96 org.apache.log4j.helpers.PatternParser
     4    32   128 org.apache.log4j.helpers.PatternParser$BasicPatternConverter
     2    40    80 org.apache.log4j.helpers.PatternParser$CategoryPatternConverter
     2    40    80 org.apache.log4j.helpers.PatternParser$DatePatternConverter
     8    32   256 org.apache.log4j.helpers.PatternParser$LiteralPatternConverter
     1    32    32 org.apache.log4j.helpers.QuietWriter
     1    16    16 org.apache.log4j.or.RendererMap
     1    40    40 org.apache.log4j.spi.RootLogger
    19    24   456 spoon.reflect.code.BinaryOperatorKind
     4    24    96 spoon.reflect.code.UnaryOperatorKind
     8    24   192 spoon.reflect.declaration.ModifierKind
     1    24    24 spoon.reflect.factory.CompilationUnitFactory
     1    80    80 spoon.reflect.factory.FactoryImpl
     1    24    24 spoon.reflect.factory.PackageFactory
     1    96    96 spoon.reflect.factory.TypeFactory
     1    40    40 spoon.reflect.visitor.DefaultJavaPrettyPrinter
     1    48    48 spoon.reflect.visitor.DefaultJavaPrettyPrinter$PrintingContext
     1    16    16 spoon.reflect.visitor.ImportScannerImpl
     1    16    16 spoon.support.DefaultCoreFactory
     1    40    40 spoon.support.JavaOutputProcessor
     1    80    80 spoon.support.StandardEnvironment
     9    56   504 spoon.support.reflect.code.CtAnnotationFieldAccesImpl
    82    56  4592 spoon.support.reflect.code.CtArrayAccessImpl
   774    56 43344 spoon.support.reflect.code.CtAssignmentImpl
  2080    56 116480 spoon.support.reflect.code.CtBinaryOperatorImpl
  4709    48 226032 spoon.support.reflect.code.CtBlockImpl
    81    48  3888 spoon.support.reflect.code.CtBreakImpl
   116    48  5568 spoon.support.reflect.code.CtCaseImpl
   104    48  4992 spoon.support.reflect.code.CtCatchImpl
   104    56  5824 spoon.support.reflect.code.CtCatchVariableImpl
    19    56  1064 spoon.support.reflect.code.CtConditionalImpl
   848    64 54272 spoon.support.reflect.code.CtConstructorCallImpl
    38    48  1824 spoon.support.reflect.code.CtContinueImpl
  3828    56 214368 spoon.support.reflect.code.CtFieldAccessImpl
   372    56 20832 spoon.support.reflect.code.CtForEachImpl
    61    56  3416 spoon.support.reflect.code.CtForImpl
  1548    56 86688 spoon.support.reflect.code.CtIfImpl
 10134    72 729648 spoon.support.reflect.code.CtInvocationImpl
  3308    48 158784 spoon.support.reflect.code.CtLiteralImpl
  1589    56 88984 spoon.support.reflect.code.CtLocalVariableImpl
    53    56  2968 spoon.support.reflect.code.CtNewArrayImpl
    24    64  1536 spoon.support.reflect.code.CtNewClassImpl
    22    64  1408 spoon.support.reflect.code.CtOperatorAssignmentImpl
  1701    48 81648 spoon.support.reflect.code.CtReturnImpl
   111    56  6216 spoon.support.reflect.code.CtSuperAccessImpl
    16    48   768 spoon.support.reflect.code.CtSwitchImpl
   738    48 35424 spoon.support.reflect.code.CtThisAccessImpl
   156    48  7488 spoon.support.reflect.code.CtThrowImpl
    93    56  5208 spoon.support.reflect.code.CtTryImpl
   354    48 16992 spoon.support.reflect.code.CtTypeAccessImpl
   311    56 17416 spoon.support.reflect.code.CtUnaryOperatorImpl
  8972    56 502432 spoon.support.reflect.code.CtVariableAccessImpl
    30    48  1440 spoon.support.reflect.code.CtWhileImpl
   318    40 12720 spoon.support.reflect.cu.CompilationUnitImpl
 49773    32 1592736 spoon.support.reflect.cu.SourcePositionImpl
   906    56 50736 spoon.support.reflect.declaration.CtAnnotationImpl
   906    56 50736 spoon.support.reflect.declaration.CtAnnotationImpl$1
     6    64   384 spoon.support.reflect.declaration.CtAnnotationTypeImpl
     2    48    96 spoon.support.reflect.declaration.CtAnonymousExecutableImpl
   221    80 17680 spoon.support.reflect.declaration.CtClassImpl
   253    64 16192 spoon.support.reflect.declaration.CtConstructorImpl
     2    24    48 spoon.support.reflect.declaration.CtElementImpl$UNMODIFIABLE_COLLECTION
     9    80   720 spoon.support.reflect.declaration.CtEnumImpl
   661    56 37016 spoon.support.reflect.declaration.CtFieldImpl
   118    64  7552 spoon.support.reflect.declaration.CtInterfaceImpl
  2999    72 215928 spoon.support.reflect.declaration.CtMethodImpl
    28    48  1344 spoon.support.reflect.declaration.CtPackageImpl
     1    48    48 spoon.support.reflect.declaration.CtPackageImpl$1
  2561    56 143416 spoon.support.reflect.declaration.CtParameterImpl
  1518    48 72864 spoon.support.reflect.reference.CtArrayTypeReferenceImpl
   137    32  4384 spoon.support.reflect.reference.CtCatchVariableReferenceImpl
 15644    40 625760 spoon.support.reflect.reference.CtExecutableReferenceImpl
  3951    32 126432 spoon.support.reflect.reference.CtFieldReferenceImpl
  4197    32 134304 spoon.support.reflect.reference.CtLocalVariableReferenceImpl
 89369    24 2144856 spoon.support.reflect.reference.CtPackageReferenceImpl
  4638    32 148416 spoon.support.reflect.reference.CtParameterReferenceImpl
 25013    48 1200624 spoon.support.reflect.reference.CtTypeParameterReferenceImpl
 92042    40 3681680 spoon.support.reflect.reference.CtTypeReferenceImpl
     1    24    24 sun.nio.cs.MS1252
     4    64   256 sun.nio.cs.SingleByte$Encoder
     4    48   192 sun.nio.cs.StreamEncoder
     1    16    16 sun.util.calendar.Gregorian
     2    96   192 sun.util.calendar.Gregorian$Date
     2    56   112 sun.util.calendar.ZoneInfo
     1    32    32 sun.util.locale.BaseLocale
 962527       35743360 (total)
```

**Comment by leventov** [comment]
New (approximate):
```
spoon.support.reflect.declaration.CtPackageImpl instance footprint:
 COUNT   AVG   SUM DESCRIPTION
     9  3674 33072 [B
 29570    45 1345160 [C
   327   523 171336 [I
     1  1952  1952 [J
     8    19   152 [Ljava.lang.Class;
 64193    32 2095800 [Ljava.lang.Object;
    13    46   600 [Ljava.lang.String;
     8    27   216 [Ljava.lang.reflect.Field;
     3    21    64 [Ljava.lang.reflect.Method;
     1    80    80 [Ljava.util.HashMap$Node;
     2    88   176 [Ljava.util.Hashtable$Entry;
     1    64    64 [Lspoon.reflect.declaration.ModifierKind;
     2    40    80 [Z
     1    24    24 com.intellij.rt.execution.junit.segments.SegmentedOutputStream
     1    24    24 java.io.BufferedOutputStream
     2    40    80 java.io.BufferedWriter
   320    32 10240 java.io.File
     2    40    80 java.io.FileDescriptor
     2    32    64 java.io.FileOutputStream
     4    24    96 java.io.OutputStreamWriter
     2    32    64 java.io.PrintStream
     2    16    32 java.lang.Boolean
    39    16   624 java.lang.Character
    12    96  1152 java.lang.Class
     7    56   392 java.lang.Class$ReflectionData
     1    24    24 java.lang.Double
    16    16   256 java.lang.Integer
     5    24   120 java.lang.Long
     4    16    64 java.lang.Object
 29557    24 709368 java.lang.String
     5    24   120 java.lang.StringBuffer
     1    16    16 java.lang.ref.ReferenceQueue$Lock
     1    32    32 java.lang.ref.ReferenceQueue$Null
     7    40   280 java.lang.ref.SoftReference
    21    72  1512 java.lang.reflect.Field
     4    88   352 java.lang.reflect.Method
     1    24    24 java.math.RoundingMode
     4    48   192 java.nio.HeapByteBuffer
     1    16    16 java.nio.charset.CodingErrorAction
     2    64   128 java.text.DateFormatSymbols
     2   144   288 java.text.DecimalFormat
     2    64   128 java.text.DecimalFormatSymbols
     2    40    80 java.text.DigitList
     2    64   128 java.text.SimpleDateFormat
 64181    24 1540344 java.util.ArrayList
   460    16  7360 java.util.Collections$SingletonSet
     1    24    24 java.util.Currency
     4    24    96 java.util.Date
     2   112   224 java.util.GregorianCalendar
   322    48 15456 java.util.HashMap
     1    32    32 java.util.HashMap$Node
   320    16  5120 java.util.HashSet
     2    48    96 java.util.Hashtable
    12    32   384 java.util.Hashtable$Entry
     1    32    32 java.util.Locale
  8436    32 269952 java.util.RegularEnumSet
     3    32    96 java.util.Stack
   452    48 21696 java.util.TreeMap
  3936    40 157440 java.util.TreeMap$Entry
    16    16   256 java.util.TreeMap$KeySet
    20    16   320 java.util.TreeMap$Values
   448    16  7168 java.util.TreeSet
     3    32    96 java.util.Vector
    12    24   288 org.apache.log4j.CategoryKey
     1    64    64 org.apache.log4j.ConsoleAppender
     1    16    16 org.apache.log4j.DefaultCategoryFactory
     1    48    48 org.apache.log4j.Hierarchy
     3    24    72 org.apache.log4j.Level
     6    40   240 org.apache.log4j.Logger
     2    32    64 org.apache.log4j.PatternLayout
     6    32   192 org.apache.log4j.ProvisionNode
     1    88    88 org.apache.log4j.RollingFileAppender
     2    16    32 org.apache.log4j.helpers.AppenderAttachableImpl
     1    40    40 org.apache.log4j.helpers.CountingQuietWriter
     2    24    48 org.apache.log4j.helpers.FormattingInfo
     2    24    48 org.apache.log4j.helpers.OnlyOnceErrorHandler
     2    48    96 org.apache.log4j.helpers.PatternParser
     4    32   128 org.apache.log4j.helpers.PatternParser$BasicPatternConverter
     2    40    80 org.apache.log4j.helpers.PatternParser$CategoryPatternConverter
     2    40    80 org.apache.log4j.helpers.PatternParser$DatePatternConverter
     8    32   256 org.apache.log4j.helpers.PatternParser$LiteralPatternConverter
     1    32    32 org.apache.log4j.helpers.QuietWriter
     1    16    16 org.apache.log4j.or.RendererMap
     1    40    40 org.apache.log4j.spi.RootLogger
    19    24   456 spoon.reflect.code.BinaryOperatorKind
     4    24    96 spoon.reflect.code.UnaryOperatorKind
    11    24   264 spoon.reflect.declaration.ModifierKind
     1    24    24 spoon.reflect.factory.CompilationUnitFactory
     1    80    80 spoon.reflect.factory.FactoryImpl
     1    24    24 spoon.reflect.factory.PackageFactory
     1    96    96 spoon.reflect.factory.TypeFactory
     1    40    40 spoon.reflect.visitor.DefaultJavaPrettyPrinter
     1    48    48 spoon.reflect.visitor.DefaultJavaPrettyPrinter$PrintingContext
     1    16    16 spoon.reflect.visitor.ImportScannerImpl
     1    16    16 spoon.support.DefaultCoreFactory
     1    40    40 spoon.support.JavaOutputProcessor
     1    80    80 spoon.support.StandardEnvironment
     9    56   504 spoon.support.reflect.code.CtAnnotationFieldAccesImpl
    82    56  4592 spoon.support.reflect.code.CtArrayAccessImpl
   768    56 43008 spoon.support.reflect.code.CtAssignmentImpl
  2131    56 119336 spoon.support.reflect.code.CtBinaryOperatorImpl
  4709    48 226032 spoon.support.reflect.code.CtBlockImpl
    81    48  3888 spoon.support.reflect.code.CtBreakImpl
   116    48  5568 spoon.support.reflect.code.CtCaseImpl
   104    48  4992 spoon.support.reflect.code.CtCatchImpl
   104    56  5824 spoon.support.reflect.code.CtCatchVariableImpl
    19    56  1064 spoon.support.reflect.code.CtConditionalImpl
   811    64 51904 spoon.support.reflect.code.CtConstructorCallImpl
    38    48  1824 spoon.support.reflect.code.CtContinueImpl
  3915    56 219240 spoon.support.reflect.code.CtFieldAccessImpl
   368    56 20608 spoon.support.reflect.code.CtForEachImpl
    61    56  3416 spoon.support.reflect.code.CtForImpl
  1547    56 86632 spoon.support.reflect.code.CtIfImpl
 10224    72 736128 spoon.support.reflect.code.CtInvocationImpl
  3352    48 160896 spoon.support.reflect.code.CtLiteralImpl
  1588    56 88928 spoon.support.reflect.code.CtLocalVariableImpl
    53    56  2968 spoon.support.reflect.code.CtNewArrayImpl
    24    64  1536 spoon.support.reflect.code.CtNewClassImpl
    22    64  1408 spoon.support.reflect.code.CtOperatorAssignmentImpl
  1706    48 81888 spoon.support.reflect.code.CtReturnImpl
   111    56  6216 spoon.support.reflect.code.CtSuperAccessImpl
    16    48   768 spoon.support.reflect.code.CtSwitchImpl
   732    48 35136 spoon.support.reflect.code.CtThisAccessImpl
   156    48  7488 spoon.support.reflect.code.CtThrowImpl
    93    56  5208 spoon.support.reflect.code.CtTryImpl
   385    48 18480 spoon.support.reflect.code.CtTypeAccessImpl
   324    56 18144 spoon.support.reflect.code.CtUnaryOperatorImpl
  8999    56 503944 spoon.support.reflect.code.CtVariableAccessImpl
    31    48  1488 spoon.support.reflect.code.CtWhileImpl
   319    40 12760 spoon.support.reflect.cu.CompilationUnitImpl
 50085    32 1602720 spoon.support.reflect.cu.SourcePositionImpl
   906    56 50736 spoon.support.reflect.declaration.CtAnnotationImpl
   906    56 50736 spoon.support.reflect.declaration.CtAnnotationImpl$1
     6    64   384 spoon.support.reflect.declaration.CtAnnotationTypeImpl
     2    48    96 spoon.support.reflect.declaration.CtAnonymousExecutableImpl
   222    80 17760 spoon.support.reflect.declaration.CtClassImpl
   254    64 16256 spoon.support.reflect.declaration.CtConstructorImpl
     2    24    48 spoon.support.reflect.declaration.CtElementImpl$UNMODIFIABLE_COLLECTION
     9    80   720 spoon.support.reflect.declaration.CtEnumImpl
   680    56 38080 spoon.support.reflect.declaration.CtFieldImpl
   118    64  7552 spoon.support.reflect.declaration.CtInterfaceImpl
  3000    72 216000 spoon.support.reflect.declaration.CtMethodImpl
    28    48  1344 spoon.support.reflect.declaration.CtPackageImpl
     1    48    48 spoon.support.reflect.declaration.CtPackageImpl$1
  2561    56 143416 spoon.support.reflect.declaration.CtParameterImpl
  1528    48 73344 spoon.support.reflect.reference.CtArrayTypeReferenceImpl
   137    32  4384 spoon.support.reflect.reference.CtCatchVariableReferenceImpl
 15719    40 628760 spoon.support.reflect.reference.CtExecutableReferenceImpl
  4038    32 129216 spoon.support.reflect.reference.CtFieldReferenceImpl
  4202    32 134464 spoon.support.reflect.reference.CtLocalVariableReferenceImpl
 89891    24 2157384 spoon.support.reflect.reference.CtPackageReferenceImpl
  4660    32 149120 spoon.support.reflect.reference.CtParameterReferenceImpl
 25272    48 1213056 spoon.support.reflect.reference.CtTypeParameterReferenceImpl
 92582    40 3703280 spoon.support.reflect.reference.CtTypeReferenceImpl
     1    24    24 sun.nio.cs.MS1252
     4    64   256 sun.nio.cs.SingleByte$Encoder
     4    48   192 sun.nio.cs.StreamEncoder
     1    16    16 sun.reflect.DelegatingMethodAccessorImpl
     1    24    24 sun.reflect.NativeMethodAccessorImpl
     1    32    32 sun.reflect.UnsafeObjectFieldAccessorImpl
     1    32    32 sun.reflect.UnsafeQualifiedObjectFieldAccessorImpl
     1    16    16 sun.util.calendar.Gregorian
     2    96   192 sun.util.calendar.Gregorian$Date
     2    56   112 sun.util.calendar.ZoneInfo
     1    32    32 sun.util.locale.BaseLocale
 542721       19225840 (total)
```

**Comment by leventov** [comment]
Use https://www.diffchecker.com/ to compare

**Comment by GerardPaligot** [comment]
What does the columns "COUNT", "AVG" and "SUM" in your tables?

**Comment by leventov** [comment]
See OpenJDK JOL tool format. Shortly speaking, before optimizations AST takes 35.7 Mb;  after - 19.2

**Comment by monperrus** [comment]
Looks convincing, thanks. We try to reproduce your measurements.

**Comment by GerardPaligot** [comment]
I didn't know that tool. It seems great but you can do a lot of things with it. I can get some metrics but not yours. Can you give us your commands to generate these tables?

**Comment by leventov** [comment]
@GerardPaligot 

```
Launcher spoon = new Launcher();
SpoonCompiler comp = spoon.createCompiler();
comp.addInputSources(SpoonResourceHelper.resources("./src/main/java/spoon"));
comp.build();
System.gc();
comp.getFactory().Package().getAllRoots().stream()
            .map(GraphLayout::parseInstance)
            .map(GraphLayout::toFootprint)
            .forEach(System.out::println);
```

**Comment by GerardPaligot** [comment]
Ok, I got results very similar of yours. I confirm your measurements and I ran this PR on our Jenkins. Everything is okay!

File: src/main/java/spoon/support/reflect/declaration/CtClassImpl.java [hunk 1]
```diff
@@ -129,10 +130,15 @@ public void addConstructor(CtConstructor<T> constructor) {
 
 	@Override
 	public void removeConstructor(CtConstructor<T> constructor) {
-		if (constructors == CtElementImpl.<CtConstructor<T>> EMPTY_SET()) {
-			constructors = new TreeSet<CtConstructor<T>>();
+		if (!constructors.isEmpty()) {
+			if (constructors.size() == 1) {
+                if (constructors.contains(constructor)) {
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/declaration/CtClassImpl.java]: problem with indentation

**Comment by leventov** [comment]
Btw I found more problems with indentation both in existing source code, and the source added by me in this PR. but my regex-fu is not strong enough to resolve them all. I suggest you to do this as a separate commit.

Now I configured Intellij to use tabs for spoon project, so I won't smap with spaces in future.

**Comment by monperrus** [comment]
Thanks for the highly technical contribution.

**Comment by monperrus** [comment]
> Now I configured Intellij to use tabs for spoon project, so I won't
> smap with spaces in future.
Perfect.

--Martin

**Pull request merged**